ens-improvement-proposals/ensip-17-evmchain-reverse-resolution.md

```diff
@@ -0,0 +1,92 @@
+---
+description: >-
+  Specifies reverse resolution in a cross-chain context
+---
+
+# ENSIP-XX: EVM-chain Reverse Resolution
+
+| **Author**    | Jeff Lau \<jeff@ens.domains> |
+| ------------- | ---------------------------- |
+| **Status**    | Draft                        |
+| **Submitted** | 2023-03-14                   |
+
+## Abstract
+
+This ENSIP specifies a way for reverse resolution to be used on other EVM chains. This allows setting primary names on L2s, as well as resolving any other records set on this specific reverse record, such as the avatar.
+
+## Motivation
+
+Reverse resolution has been in use since ENS's inception, however at the time Ethereum had no concrete scaling plans. In the past 5 years, we've seen layer 2s and sidechains become more prevalent and we first allowed support for these with ENSIP-9 (formerly EIP-2304) to allow addresses from other chains to be stored on ENS. Reverse resolution can be expanded to allow the forward resolution to first check the records for the chain that the user is on.
+
+
+## Specification
+
+### Overview
+
+* Reverse registrars will be setup on each L2, with a corresponding registry
+* Reverse registrar will only allow setting the name, without resolver customisability. This is to allow a consistent storage slot that can be checked on L1.
+* User can now claim their reverse on this L2 and set it to their ENS name
+* Their ENS name will need to set their record for the same EVM-cointype as the network, which is specified in [ENSIP-11](https://docs.ens.domains/ensip/11).
+* A dapp can then detect the chainID that a user is on, find the corresponding cointype and resolve their primary ens name by resolving the name record at [userAddress].[coinType].reverse. This will be resolved via CCIP-read and look up the reverse record on the corresponding EVM-chain.
+* Dapp will then resolve this name via ENS on L1 to check if the forward resolution matches. This forward resolution can be on L1, or the user can set up CCIP-read records for each network and put those addresses wherever they want.
+* Once matched, the dapp can then also resolve any text records on the primary ENS name, such as avatar.
+
+### Resolving Primary ENS Names by a dapp
+
+1) If a dapp has a connected user it SHOULD first check the chainId of the user.
+2) If the chainId is not 1, it SHOULD then construct the ENS name [address].[coinType].reverse to obtain the primary ENS name of the user. 
+3) If none is found, it SHOULD check check mainnet [address].default.reverse.
```

Comment:
We should clarify behaviour here if `[address].[cointype].reverse` has no resolver vs if it does but does not implement `name`, or returns the empty string.

Reply:
I believe we can't have no resolver in our current architecture because we're putting the L2 reverse registrar together with the name/text records

Reply:
At a minimum this needs to be clearer than "if none is found". In general this needs to be much more specific and actionable - describe the specific methods and calls that need to be made and what checks should be done on the return values. When that's already specified elsewhere - eg, forward resolution - reference the relevant standard.

Reply:
I've expanded on this considerably to make it more explicit